src/vaults/harvests.utils.ts

```diff
@@ -1,164 +1,271 @@
+  return timestamps.map((t) => {
+    const timestamp = Number(t);
+    const currentHarvests = harvestsByTimestamp.get(timestamp) ?? [];
+    const currentDistributions = treeDistributionsByTimestamp.get(timestamp) ?? [];
+    return {
+      timestamp,
+      harvests: currentHarvests.map((h) => ({
+        timestamp,
+        block: Number(h.blockNumber),
+        token: vault.depositToken,
+        amount: h.amount,
+      })),
+      treeDistributions: currentDistributions.map((d) => {
+        const tokenAddress = d.token.id.startsWith('0x0x') ? d.token.id.slice(2) : d.token.id;
+        return {
+          timestamp,
+          block: Number(d.blockNumber),
+          token: ethers.utils.getAddress(tokenAddress),
+          amount: d.amount,
+        };
+      }),
+    };
+  });
+}
 
-  const fmtBalance = formatBalance(balance, depositToken.decimals);
+function constructYieldItems(data: VaultHarvestData[]): VaultYieldItem[] {
+  const recentHarvests = data.sort((a, b) => b.timestamp - a.timestamp);
+  const allHarvests = recentHarvests.flatMap((h) => h.harvests.map((h) => ({ ...h, type: YieldType.Harvest })));
+  const allDistributions = recentHarvests.flatMap((h) =>
+    h.treeDistributions.map((d) => ({ ...d, type: YieldType.Distribution })),
+  );
+  return allHarvests
+    .concat(allDistributions)
+    .filter((e) => BigNumber.from(e.amount).gt(ethers.constants.Zero))
+    .sort((a, b) => b.timestamp - a.timestamp);
+}
 
-  const totalHarvestedTokens = formatBalance(amount || BigNumber.from(0), depositToken.decimals);
-  const durationScalar = ONE_YEAR_SECONDS / duration;
+async function getVaultBalance(chain: Chain, vault: VaultDefinitionModel, block: number): Promise<number> {
+  const sdk = await chain.getSdk();
 
-  const compoundApr = (totalHarvestedTokens / fmtBalance) * durationScalar * 100;
+  let balance = 0;
+  if (isInfluenceVault(vault.address)) {
+    balance = await getInfuelnceVaultYieldBalance(chain, vault, block);
+  } else {
+    const vaultContract = Vault__factory.connect(vault.address, sdk.provider);
+    const totalSupply = await vaultContract.totalSupply({ blockTag: block });
+    balance = formatBalance(totalSupply);
+  }
 
-  return parseFloat(compoundApr.toFixed(2));
+  return balance;
 }
 
-export async function getVaultHarvestsOnChain(
+/**
+ * Evaluate yield events to derivate base apr, and apy as well as apr of any emitted tokens.
+ * @param chain network the vault is deployed on
+ * @param vault vault requested yield event evaluation
+ * @param yieldEvents events sourced from either on chain or the graph
+ * @returns yield summary providing the base information for construction of yield sources
+ */
+async function evaluateYieldItems(
   chain: Chain,
-  address: VaultDefinitionModel['address'],
-  startFromBlock: Nullable<number> = null,
-): Promise<VaultHarvestsExtendedResp[]> {
-  const vaultHarvests: VaultHarvestsExtendedResp[] = [];
+  vault: VaultDefinitionModel,
+  yieldItems: VaultYieldItem[],
```

Comment:
would be good to have check that yieldItems are in range

Reply:
looking back, this one is for the any items not just the tway range so i think it is not needed